### PR TITLE
document coding conventions for templates

### PR DESCRIPTION
**Pull request by davydden**
fixes https://github.com/dealii/dealii/issues/3291  (mostly a copy-paste from comments).

### CODE REVIEW

File: doc/doxygen/headers/coding_conventions.h [hunk 1]
```diff
@@ -161,6 +161,47 @@ source files.
...
+types -- because users might want to define their own vector kinds) but at
+least know a few common usage cases, then the function is put into a
+<code>.templates.h</code> file. We #include it into the <code>.cc</code> file and instantiate the
+functions for all of the common arguments. For almost all users, this will be
+just fine -- they only use the (vector, matrix, ...) types we already
+instantiate, and for them the <code>.templates.h</code> file will not be of any interest.
+It will also not slow down their compilations because nothing they see will
+#include the <code>.templates.h</code> file. But users who define their own
+(vector, matrix, ...) types can instantiate the template functions with their
+own user-defined types by #includeing the <code>.templates.h</code> files.
```
Comment (tjhei) [inline doc/doxygen/headers/coding_conventions.h]: just use "including"

File: doc/doxygen/headers/coding_conventions.h [hunk 2]
```diff
@@ -161,6 +161,47 @@ source files.
...
+<li> Finally, if we can not assume in advance which values template arguments
+will take (e.g., any class derived from Subscriptor can be used as an argument),
+the definitions of functions are provided at the bottom of the header
+file with declarations. The definitions should be guarded with <code>#ifndef DOXYGEN ...
+#endif</code> to prevent Doxygen from picking them up.</li>
+
+</ol>
+
+<p> For the first two cases, instantiation instructions are defined in <code>.inst.in</code>
+files. These files are processed by custom CMake scripts to generate <code>.inst</code> files
```
Comment (tjhei) [inline doc/doxygen/headers/coding_conventions.h]: They are processed by a binary called ``expand_instantiations`` (built from cmake/scripts/expand_instantiations.cc) and the parameters are defined dynamically through cmake depending on your configuration (see `` share/deal.II/template-arguments`` in your build directory).

**Comment by davydden** [comment]
@tjhei i updated the text.

**Comment by Rombur** [comment]
:+1:

**Pull request merged**